llvm/lib/Transforms/Vectorize/VectorCombine.cpp

```diff
@@ -2988,6 +2989,241 @@ bool VectorCombine::foldShuffleFromReductions(Instruction &I) {
   return foldSelectShuffle(*Shuffle, true);
 }
 
+bool VectorCombine::foldShuffleChainsToReduce(Instruction &I) {
+  auto *EEI = dyn_cast<ExtractElementInst>(&I);
+  if (!EEI)
+    return false;
+
+  std::queue<Value *> InstWorklist;
+  InstructionCost OrigCost = 0;
+
+  Value *InitEEV = nullptr;
+
+  unsigned int CommonCallOp = 0;
+  Instruction::BinaryOps CommonBinOp = Instruction::BinaryOpsEnd;
+
+  bool IsFirstCallOrBinInst = true;
+  bool ShouldBeCallOrBinInst = true;
+
+  SmallVector<Value *, 3> PrevVecV(3, nullptr);
```

Comment:
I cannot understand what this array is used for. You always access its elements with constant indices, so these elements seem to have different meanings. Can you add some comments to explain the meaning of each element?

Reply:
Yeah, documentation was the last thing left to add for this fold pattern. Since you asked, `PrevVecV` is used to store the last two instructions which are used either singly for shuffle vector or for binary operations. Specifically, `0` and `1` would be the ones that keep changing as we go up, whereas `2` is the one we get from extract operation.

These help differentiate between them in case they are mixed up.